opt(stream): add option to directly copy over tables from lower levels (#1700)

mangalaman93 opened the pull request:
Also takes a bug fix from PR https://github.com/dgraph-io/badger/pull/1712, commit https://github.com/dgraph-io/badger/commit/58d06744496364160eb50722fe171894fb139619

This PR adds FullCopy option in Stream. This allows sending the table entirely to the writer. If this option is set to true we directly copy over the tables from the last 2 levels. This option increases the stream speed while also lowering the memory consumption on the DB that is streaming the KVs.

For 71GB, compressed and encrypted DB we observed 3x improvement in speed. The DB contained ~65GB in the last 2 levels while remaining in the above levels.

To use this option, the following options should be set in Stream.

stream.KeyToList = nil
stream.ChooseKey = nil
stream.SinceTs = 0
db.managedTxns = true

If we use stream writer for receiving the KVs, the encryption mode has to be the same in sender and receiver. This will restrict db.StreamDB() to use the same encryption mode in both input and output DB. Added TODO for allowing different encryption modes.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dgraph-io/badger?pullRequest=1872) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dgraph-io/badger?pullRequest=1872) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dgraph-io/badger?pullRequest=1872) it.</sub>

mangalaman93 committed:
opt(stream): add option to directly copy over tables from lower levels (#1700)

Also takes a bug fix from PR #1712, commit 58d0674

This PR adds FullCopy option in Stream. This allows sending the
table entirely to the writer. If this option is set to true we
directly copy over the tables from the last 2 levels. This option
increases the stream speed while also lowering the memory
consumption on the DB that is streaming the KVs.

For 71GB, compressed and encrypted DB we observed 3x improvement
in speed. The DB contained ~65GB in the last 2 levels while
remaining in the above levels.

To use this option, the following options should be set in Stream.

stream.KeyToList = nil
stream.ChooseKey = nil
stream.SinceTs = 0
db.managedTxns = true

If we use stream writer for receiving the KVs, the encryption mode
has to be the same in sender and receiver. This will restrict
db.StreamDB() to use the same encryption mode in both input and
output DB. Added TODO for allowing different encryption modes.

joshua-goldstein commented on stream_writer.go:
```diff
@@ -109,7 +117,64 @@ func (sw *StreamWriter) Write(buf *z.Buffer) error {
+			// the prevLevel, we know we're processing a last level table.
+			// The last level for this DB can be 8 while the DB that's sending
+			// this could have the last level at 7.
+			if sw.senderPrevLevel != level {
```
can we test streaming between two instances of badger with different number of levels (I.e. smaller to bigger, bigger to smaller)?

joshua-goldstein replied:
also, can we add some guardrails w.r.t. checking for the potential size of new database and number of levels?

joshua-goldstein commented on level_handler.go:
```diff
@@ -324,14 +324,41 @@ func (s *levelHandler) appendIterators(iters []y.Iterator, opt *IteratorOptions)
+		panic("Invalid option for getTables")
+	}
+
+	// Typically this would only be called for the last level.
```
in this case we are calling this on all levels correct?

mangalaman93 replied:
Done

mangalaman93 commented:
This code has a data race, I am looking into it:
```
==================
WARNING: DATA RACE
Write at 0x00c000b100e8 by goroutine 2957:
  github.com/dgraph-io/badger/v3.(*StreamWriter).Write.func1()
      /home/aman/gocode/pkg/mod/github.com/dgraph-io/badger/v3@v3.2103.6-0.20230214155941-0e7c6a7a614a/stream_writer.go:212 +0x167
  github.com/dgraph-io/ristretto/z.(*Buffer).SliceIterate()
      /home/aman/gocode/pkg/mod/github.com/dgraph-io/ristretto@v0.1.1/z/buffer.go:290 +0x30a
  github.com/dgraph-io/badger/v3.(*StreamWriter).Write()
      /home/aman/gocode/pkg/mod/github.com/dgraph-io/badger/v3@v3.2103.6-0.20230214155941-0e7c6a7a614a/stream_writer.go:146 +0x194
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*countIndexer).writeIndex()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/count_index.go:188 +0x9c7
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*countIndexer).addCountEntry.func1()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/count_index.go:100 +0x47

Previous write at 0x00c000b100e8 by goroutine 228:
  github.com/dgraph-io/badger/v3.(*StreamWriter).Write.func1()
      /home/aman/gocode/pkg/mod/github.com/dgraph-io/badger/v3@v3.2103.6-0.20230214155941-0e7c6a7a614a/stream_writer.go:212 +0x167
  github.com/dgraph-io/ristretto/z.(*Buffer).SliceIterate()
      /home/aman/gocode/pkg/mod/github.com/dgraph-io/ristretto@v0.1.1/z/buffer.go:290 +0x30a
  github.com/dgraph-io/badger/v3.(*StreamWriter).Write()
      /home/aman/gocode/pkg/mod/github.com/dgraph-io/badger/v3@v3.2103.6-0.20230214155941-0e7c6a7a614a/stream_writer.go:146 +0x194
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).startWriting.func2()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:349 +0xd1
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).startWriting()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:389 +0x284
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).reduce.func4()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:475 +0x64

Goroutine 2957 (running) created at:
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*countIndexer).addCountEntry()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/count_index.go:100 +0x4c9
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).startWriting.func1.2()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:338 +0x4e
  github.com/dgraph-io/ristretto/z.(*Buffer).SliceIterate()
      /home/aman/gocode/pkg/mod/github.com/dgraph-io/ristretto@v0.1.1/z/buffer.go:290 +0x30a
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).startWriting.func1()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:336 +0x1cb
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).startWriting()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:392 +0x2bc
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).reduce.func4()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:475 +0x64

Goroutine 228 (running) created at:
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).reduce()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:475 +0x4c5
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).run.func1()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:109 +0xbe4
  github.com/dgraph-io/dgraph/dgraph/cmd/bulk.(*reducer).run.func2()
      /home/aman/gocode/src/github.com/dgraph-io/dgraph/dgraph/cmd/bulk/reduce.go:123 +0x66
==================
```

harshil-goel reviewed (commented):
Can you write a benchmark for this?

ryanfoxtyler commented:
closing PR for now, keeping branch active
